### DISCUSSION TITLE
Nested Mesh Interaction

### DISCUSSION THREAD

**Discussion by klaw13** [idaholab/moose#18267] [Q&A Modules: Navier-Stokes] [answered]
Hi there, 

I'm trying to simulate the flow of water inside a cavity flowing inside the red volume. As the water flows throughout the cavity, the pressure from the flow should be able to push on the mesh boundary to move it. If the force is large 
enough, both the outer and inner volumes should become cylinders. 

![Screenshot from 2021-07-06 08-47-07](https://user-images.githubusercontent.com/20307070/124611358-fb42ee80-de36-11eb-9562-4df0b98512b2.png)

input file: [input.i.txt](https://github.com/idaholab/moose/files/6770374/input.i.txt) )

I initially tried to just simulate the flow of the water inside the cavity without considering any deformations to the mesh. When I run the simulation, I can get the water to flow about half way through, but then the simulation fails to converge. 

I thought it might be because I need to take into consideration the effects on the mesh, so I  tried to base my approach on the 3D_struc_acoustic.i example, and added in kernels to handle the convected mesh. 

However, I have not been able to converge at any timestep. Currently, I'm getting the following error: 

`*** ERROR ***
MooseVariableFE: Old time derivative of solution (`u_dot_old`) is not stored. Please set uDotOldRequested() to true in FEProblemBase before requesting `u_dot_old`.
`

**Comment by GiudGiud** [comment]
Hello

The effect on the mesh should be a sufficiently different issue that you should try to get a resolved solution without modeling this effect first. 

Once you get a good solution without the fluid structure interaction, @somu15 should be able to help with that error message, which I assume is related to a FSI kernel

Best,
Guillaume

**Reply by somu15** [reply]
Hi @klaw13. As @GiudGiud suggested, try and resolve the problem without FSI. With the FSI, it is important how you define the element and neighbor boundaries when setting the sidesets. I will provide more information once you let me know you have resolved your issue without FSI. Thanks!

**Reply by klaw13** [reply]
Thanks @GiudGiud. 

Hi @somu15

So my convergence issue seems to be due to the values I have set in the parameters of my executioner block  - so establishing the flow of water was not my issue. However, I'm still running into the same error when I try to incorporate FSI with the mesh:

*** ERROR *** MooseVariableFE: Old time derivative of solution (u_dot_old) is not stored. Please set uDotOldRequested() to true in FEProblemBase before requesting u_dot_old. 

Taking a look at a simpler problem, I tried to modify the fsi_flat_channel.i example within the MOOSE repo: [simplified.i.txt](https://github.com/idaholab/moose/files/6812628/simplified.i.txt)

I added the following to the kernels  
- [./diffusion]
    type = Diffusion
    variable = 'p'
    block = 0
  [../]
- [./inertia]
    type = AcousticInertia
    variable = p
    block = 0
  [../]
- [./DynamicTensorMechanics]
    displacements = 'disp_x disp_y'
    block = 1
  [../]

and then changed the InterfaceKernels block from using CoupledPenaltyInterfaceDiffusion to instead use the StructureAcousticInterface module. This input file setup also results in the same error regardless of the interface kernel I choose.

**Reply by GiudGiud** [reply]
Do you have a working fluid-flow only simulation now?

@somu15 for the variable error

**Reply by klaw13** [reply]
Yes, I have a working fluid-flow only simulation now.

**Reply by somu15** [reply]
Hi @klaw13 . Could you try turning off the INS kernels and commenting out the velocity variables for the fluid domain and run FSI?

**Reply by klaw13** [reply]
Hi @somu15, it runs successfully when INS kernels are turned off and velocity variables are removed.

**Reply by somu15** [reply]
@klaw13 That is what I had expected. The FSI you are using is an acoustic FSI where pressure is the only primary variable in the fluid domain. See the physics here: https://mooseframework.inl.gov/modules/fsi/fsi_acoustics.html I believe @fdkong has an interface kernel that can work with INS kernels. Thank you.

**Reply by klaw13** [reply]
Thanks @somu15.

Also, are there any specific requirements of how the mesh should be defined (more specifically at the interface) for FSI to work correctly?

**Reply by somu15** [reply]
@klaw13  Yes, when defining the interface either in Cubit or MOOSE ('SideSetsBetweenSubdomainsGenerator'), the primary side should be the fluid and secondary side should be the structure. This way, the normal vector would point from the fluid to the structure. Correspondingly, in the 'StructureAcousticInterface', the primary variable should be 'p' and neighbor variable should be 'disp'.

**Reply by klaw13** [reply]
Hi @somu15 I tried using  'StructureAcousticInterface' for this geometry without any INS kernels. I adjusted the 3D_struc_acoustic.i example for my problem, but still receive the error as before: 
*** ERROR ***
MooseVariableFE: Old time derivative of solution (`u_dot_old`) is not stored. Please set uDotOldRequested() to true in FEProblemBase before requesting `u_dot_old`.

I've also verified with the geometry that the blocks of the solid and liquid regions are merged together and display in the mesh output, and that the primary variable is 'p' and neighbor variables are 'disp' as you stated.

**Reply by somu15** [reply]
Hi @klaw13 . I will look into this issue and get back to you. A small clarification: while generating the sideset for the interface between fluid and structure, you set the primary_side as fluid and secondary side as solid, correct? If you are generating the mesh using MOOSE mesh generators, you can directly do this. If you are using CUBIT, you need use the 'wrt' flag while defining the interface sideset. I'm confident you have already done this, but just wanted to make sure.
Thanks!

**Reply by klaw13** [reply]
@somu15 Yes, I've used the mesh generator and specified the primary side as the fluid and solid as secondary. Thanks for looking into this!

**Reply by somu15** [reply]
Sure. Would it be possible to share your recent input file and cubit exodus file if any? Thanks.

**Reply by klaw13** [reply]
Sure I've attached my current input file and output of the mesh

[mesh_output.zip](https://github.com/idaholab/moose/files/7287168/mesh_output.zip
[input.i.txt](https://github.com/idaholab/moose/files/7287169/input.i.txt)
)

**Reply by somu15** [reply]
@klaw13 I am able to run the transient simulation. The biggest change is using the `NewmarkBeta` time integrator instead of `Heun`. I don't know why we can't work with `Heun` as I have never used this before. I am attaching the input file that I have used. I am also attaching an image showing that the simulation is running. A couple of suggestions:

1. You can use the `DynamicTensorMechanics` instead of setting up the `ADStressDivergenceTensors` kernels manually. With this, you don't have to provide as much information to the `InertialForce` kernel as you are doing previously.
2. Try using `ComputeSmallStrain` in the `Materials` block if this is justifiable. Then, you can use `Newton` in the solve_type. Right now, the solves are not converging well but that is a different issue. :)

<img width="1424" alt="Screen Shot 2021-10-05 at 10 10 32 AM" src="https://user-images.githubusercontent.com/19966054/136061388-efee8cab-be38-4b7f-be02-7a6b9827838b.png">

[input copy.i.txt](https://github.com/idaholab/moose/files/7287390/input.copy.i.txt)
.

**Reply by klaw13** [reply]
Thanks @somu15! Making those changes works. That is quite strange it doesn't work with the Heun time integrator.

**Comment by fdkong** [comment, marked as answer]
It is a fascinating application. We currently do not have a 3D example in FSI module, but we have a 2D model, "fsi/test/tests/fsi_2d", with which you could start. 

The example is simple, but it is pretty general. Most of the components, internals, and kernels, should be ready to use for your applications. 

> Also, are there any specific requirements of how the mesh should be defined (more specifically at the interface) for FSI to work correctly?

We require solid and fluid meshes to match each other on the interface, even though that can be lifted in the future.

**Reply by klaw13** [reply]
Thanks @fdkong and @somu15! 

More on the mesh specifications, are Cubit and the MOOSE mesh generators the only softwares currently compatible with the FSI module? I am using gmsh which doesn't seem to have an option to specify the primary and secondary sides of an interface. 

I was able to extend the fsi_2d example into 3d with the MOOSE mesh, but when I recreate it with gmsh while ensuring that the meshes match on each side of the interface, the simulation fails to implement the FSI module correctly.

**Reply by fdkong** [reply]
> More on the mesh specifications, are Cubit and the MOOSE mesh generators the only softwares currently compatible with the FSI module? I am using gmsh which doesn't seem to have an option to specify the primary and secondary sides of an interface.

Hmm, no, we do not put any restriction on FSI module. I have minimal experience with gmsh, but we have a lot of users who use gmsh. 

You do not necessarily need to generate or mark primary and secondary sides interface using gmsh.  You only need to create fluid and solid meshes using gmsh. Of course, you need to mark each of the domains as a block. And then, you could use `SideSetsBetweenSubdomainsGenerator` (as @somu15  suggested earlier) to generate interface side sets.

> but when I recreate it with gmsh while ensuring that the meshes match on each side of the interface, the simulation fails to implement the FSI module correctly.

What were error messages?

**Reply by klaw13** [reply]
I wasn't getting an explicit error message, but the fluid would go through the fluid block and nothing would happen to the solid block which is not what we would expect. 

I made two rectangular blocks and specified them accordingly in gmsh, but when I use the SideSetsBetweenSubdomainsGenerator and run the simulation, I get the error that the side set id does not exist...

My mesh definition syntax was 

[Mesh]
   [main]
      type = FileMeshGenerator
      file = '1side_v0.msh'
   []
   [interface]
      type = SideSetsBetweeSubdomainsGenerator
      input = main
      primary_block = 'fluid'
      paired_block = 'solid'
      new_boundary = 'in_between'
    [] 
[] 

And in my InterfaceKernels block, I reference the boundary 'in_between' Which based on the examples should be correct.

**Reply by fdkong** [reply]
Run your code with "--mesh-only", and then use paraview to check if blocks "fluid" and "solid", and interface set  "in_between" do exist

**Reply by klaw13** [reply]
When running "--mesh-only", the solid and fluid blocks exist. However, the "in_between" interface set does not exist.

**Reply by fdkong** [reply]
Do you mind sharing the mesh with us?

**Reply by klaw13** [reply]
Sure! [1side_v0.zip](https://github.com/idaholab/moose/files/6920028/1side_v0.zip)

**Reply by fdkong** [reply]
I meant the exodus file that was  generated by using "--mesh-only"

**Reply by klaw13** [reply]
[mesh_exodus.zip](https://github.com/idaholab/moose/files/6920151/mesh_exodus.zip)

**Reply by fdkong** [reply]
Does gmsh has an equivalent command-line of "merge volume solid fluid" of cubit?

The interface is not merged between fluid and solid. That can be done via "merge volume solid fluid" in cubit. Could you do the same thing in gmsh?

**Reply by GiudGiud** [reply]
@klaw13 any updates on this?

**Reply by klaw13** [reply]
@fdkong Yes, fortunately I was able to figure out how to merge the volumes together directly in gmsh. The simulation runs now with the extended fsi_3d version. I'm still working on applying this setup to my custom application and will follow up if I continue getting errors. Thank you!
